/**
 * Autogenerated by Avro
 * 
 * DO NOT EDIT DIRECTLY
 */
package org.opencb.hpg.bigdata.app.cli.local.avro;
@SuppressWarnings("all")
/** Evidence of pathogenicity and benign impact as defined in Richards, S. et al. (2015). Standards and guidelines for the interpretation
    of sequence variants: a joint consensus recommendation of the American College of Medical Genetics and Genomics and
    the Association for Molecular Pathology. Genetics in Medicine, 17(5), 405–423. https://doi.org/10.1038/gim.2015.30

Evidence of pathogenicity:
* `very_strong`:
    - PVS1 null variant (nonsense, frameshift, canonical ±1 or 2 splice sites, initiation codon, single or multiexon
    deletion) in a gene where LOF is a known mechanism of disease
* `strong`:
    - PS1 Same amino acid change as a previously established pathogenic variant regardless of nucleotide change
    - PS2 De novo (both maternity and paternity confirmed) in a patient with the disease and no family history
    - PS3 Well-established in vitro or in vivo functional studies supportive of a damaging effect on the gene or gene
    product
    - PS4 The prevalence of the variant in affected individuals is significantly increased compared with the prevalence
    in controls
* `moderate`:
    - PM1 Located in a mutational hot spot and/or critical and well-established functional domain (e.g., active site of
    an enzyme) without benign variation
    - PM2 Absent from controls (or at extremely low frequency if recessive) in Exome Sequencing Project, 1000 Genomes
    Project, or Exome Aggregation Consortium
    - PM3 For recessive disorders, detected in trans with a pathogenic variant
    - PM4 Protein length changes as a result of in-frame deletions/insertions in a nonrepeat region or stop-loss
    variants
    - PM5 Novel missense change at an amino acid residue where a different missense change determined to be pathogenic
    has been seen before
    - PM6 Assumed de novo, but without confirmation of paternity and maternity
* `supporting`:
    - PP1 Cosegregation with disease in multiple affected family members in a gene definitively known to cause the
    disease
    - PP2 Missense variant in a gene that has a low rate of benign missense variation and in which missense variants are
    a common mechanism of disease
    - PP3 Multiple lines of computational evidence support a deleterious effect on the gene or gene product
    (conservation, evolutionary, splicing impact, etc.)
    - PP4 Patient’s phenotype or family history is highly specific for a disease with a single genetic etiology
    - PP5 Reputable source recently reports variant as pathogenic, but the evidence is not available to the laboratory
    to perform an independent evaluation

Evidence of benign impact:
* `stand_alone`:
    - BA1 Allele frequency is >5% in Exome Sequencing Project, 1000 Genomes Project, or Exome Aggregation
    Consortium
* `strong`:
    - BS1 Allele frequency is greater than expected for disorder
    - BS2 Observed in a healthy adult individual for a recessive (homozygous), dominant (heterozygous), or X-linked
    (hemizygous) disorder, with full penetrance expected at an early age
    - BS3 Well-established in vitro or in vivo functional studies show no damaging effect on protein function or
    splicing
    - BS4 Lack of segregation in affected members of a family
* `supporting`:
    - BP1 Missense variant in a gene for which primarily truncating variants are known to cause disease
    - BP2 Observed in trans with a pathogenic variant for a fully penetrant dominant gene/disorder or observed in cis
    with a pathogenic variant in any inheritance pattern
    - BP3 In-frame deletions/insertions in a repetitive region without a known function
    - BP4 Multiple lines of computational evidence suggest no impact on gene or gene product (conservation,
    evolutionary, splicing impact, etc.)
    - BP5 Variant found in a case with an alternate molecular basis for disease
    - BP6 Reputable source recently reports variant as benign, but the evidence is not available to the laboratory to
    perform an independent evaluation
    - BP7 A synonymous (silent) variant for which splicing prediction algorithms predict no impact to the splice
    consensus sequence nor the creation of a new splice site AND the nucleotide is not highly conserved */
public enum EvidenceImpact {
  very_strong, strong, moderate, supporting, stand_alone  ;
}
